### ISSUE TITLE
qiskit/visualization/bloch.py needs testing

### ISSUE THREAD

**Issue by 1ucian0** [Qiskit/qiskit-terra#2304]
The code in `qiskit/visualization/bloch.py` needs testing. Current coverage is:

file | statements | missing | coverage
-- | -- | -- | --
qiskit/visualization/bloch.py | 275 | 275 | 0%

**Comment by 1ucian0** [comment]
Still true: https://coveralls.io/builds/24960860/source?filename=qiskit/visualization/bloch.py

**Comment by 1ucian0** [comment]
`on hold` until #2968 is decided.

**Comment by 1ucian0** [comment]
Removing `on hold`. The testing should be on binder, like in https://github.com/Qiskit/qiskit-terra/pull/4544

**Comment by mlarocca** [comment]
@1ucian0 do you think I could help with this one?

**Comment by 1ucian0** [comment]
Yes! Probably following the same style than in https://github.com/Qiskit/qiskit-terra/pull/4544

**Comment by mlarocca** [comment]
Sounds good, thanks! I'll give it a try

**Comment by javabster** [comment]
Hi @mlarocca are you still working on this? There are more details [here](https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#snapshot-testing-for-visualizations) about how to run snapshot tests for visualisations. And you can check coverage for a specific file by running `coverage3 report -m <file path>`. Let us know if you have any questions 😄

**Comment by a-freakish** [comment]
I am new to the world of quantum computing and open source. I have good familiarity with python. I would love to work on this issue.

**Comment by javabster** [comment]
Hi @a-freakish as @mlarocca hasn't replied I will re-assign this to you, let us know if you have any questions 😄

**Comment by mlarocca** [comment]
Hello @javabster, sorry for the late reply!
I hadn't had a chance to make a lot of progress on this back in September, and then for some reason we forgot about this (I think in favor of other tasks).
It was an interesting one and I'd have loved to pick it up again, but at the same time I'm swamped right now and for a few more weeks, so I also meant to say that if @a-freakish has the time to tackle this right now, it makes sense I step aside.

If later or at any point you need help, please feel free to let me know!

Thanks!

**Comment by javabster** [comment]
Thanks for the update @mlarocca! 😃

**Comment by mrvee-qC** [comment]
# Update for Santa:elf:- QGoGP

✅ bloch.py still lacks its own dedicated testing and has no-one assigned to it - Issue exists

As I understand, the functions of bloch.py are directly used in `plot_bloch_vector`,`plot_bloch_multivector`,`plot_state_qsphere` and `visualize_transition` including any functions that depend on any of these 4 functions

`plot_state_qsphere` and `plot_bloch_multivector` are tested in test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py

Will update comment/add more comments if cause is found/fixed!

> Python version 3.9.7
> qiskit-terra version: 0.19.1

**Comment by 1ucian0** [comment]
According with coverall, [the file is 49% covered](https://coveralls.io/builds/44969725/source?filename=qiskit%2Fvisualization%2Fbloch.py).  Still some testing is needed, but much better than the original situation.

**Comment by 1ucian0** [comment]
did you have any chance to have a look @a-freakish ?

**Comment by 1ucian0** [comment]
unassigning to return the isssue to the pool of bugs for new comers.

**Comment by orangekame3** [comment]
@1ucian0 
Hi, if nobody has been assigned to this task, I'm interested in it.

**Comment by 1ucian0** [comment]
Assigned @orangekame3! Thanks!

**Comment by orangekame3** [comment]
record the current coverage for reference.

`tox -ecoverage && .tox/coverage/bin/coverage3 html`

| Name| Stmts | Miss|Cover |
|--------|--------|--------|--------|
|qiskit/visualization/bloch.py     |  286   | 145 | 49%|

**Comment by sjbokhari** [comment]
@1ucian0 Hi is there still help needed with this one? :) I would really like to contribute.

**Comment by orangekame3** [comment]
I haven't made any progress on this, so if it's okay, I'd like to pass it on to @sjbokhari .

**Comment by 1ucian0** [comment]
reassigning. Thanks both!

**Comment by aabennak** [comment]
I would like to contribute on this issue if no progress has been done.

**Comment by sjbokhari** [comment]
@aabennak I am currently working on it, but if you want to you can take over :)

**Comment by aabennak** [comment]
@sjbokhari Thank you so much for your kind reply :) but I can't do that if you're currently still working on it. Good luck!

edit: I understand that the learning curve is a thing, so me butting in would destroy that time that you have invested in understanding the code.

**Comment by 1ucian0** [comment]
any news @sjbokhari ?

**Comment by sjbokhari** [comment]
Hey @1ucian0 I am still on it I will give an update asap :) Thank you for your patience

**Comment by KynesIam** [comment]
I am new to open source code work and this seems like a good starting place.  Is anyone currently working on this issue or could it be assigned to me?

**Comment by jstowers** [comment]
Hi @1ucian0,

I am also new to open source.  Writing unit tests is a good way to learn the codebase.  I will investigate.

**Comment by 1ucian0** [comment]
@jstowers assigned.

@KynesIam , if you already have a PR ready, please submitted.

**Comment by 1ucian0** [comment]
The current coverage for `/qiskit/visualization/bloch.py` is 49.66%

Cleaning assignees as it seems stalling.

**Comment by dreads** [comment]
@1ucian0 the coverage in https://github.com/Qiskit/qiskit/pull/16738 was 99% in my local tests.

**Comment by 1ucian0** [comment]
@dreads, your lengthy, unsupervised LLM-based explanations in issues and PRs are not very helpful. Please be considerate of the humans who spend real human time reading them.

**Comment by 1ucian0** [comment]
`qiskit.visualization.bloch` hosts internal classes and functions. For extending coverage of that code means improving the testing of public API `plot_bloch_multivector` or `plot_bloch_vector`.

**Comment by dreads** [comment]
You have my apologies, @1ucian0. What I took as being thorough was unhelpful.

**Comment by dreads** [comment]
@1ucian0 I've added the image regression tests and reference images and updated the base branch in the PR. If you would prefer, I can go further than pixel-diffs and inspect the returned Figure's artists directly, asserting exact numeric/text correctness.
